feat(resource) Introduce remote_file

holyhope opened the pull request:
Hi @NikolaLohinski 👋

Fix #5.

First version includes:
- [x] `source_url`, `destination_path` and `checksum`, credentials attributes.
- [x] Create a download task when creating or updating a resource.
- [x] Delete files before updating or deleting a resource.
- [x] Import a remote file resource from download task id.
- [x] Stop the download task before updating or deleting a resource. Requires https://github.com/NikolaLohinski/free-go/pull/2
- [x] Delete the download task before updating or deleting a resource. Requires https://github.com/NikolaLohinski/free-go/pull/1
- [x] Import a remote file resource from a path.
- [x] Use [file hash Freebox API](https://dev.freebox.fr/sdk/os/fs/#hash-a-file) to read and verify resource checksum: https://github.com/NikolaLohinski/free-go/pull/4
- [x] Return an error when the file already exist.

Not implemented:

- Import basic auth from a download task: The API does not expose this information.
- Create a remote_file from a local one: The `free-go` sdk does not implement yet the upload feature.
- Create a remote_file from content: The `free-go` sdk does not implement yet the upload feature.

Note:
To test the remote file go tests, I used `https://raw.githubusercontent.com/holyhope/terraform-provider-freebox/refs/heads/resources/remote_file/examples/provider.freebox.tf` as example source:
```diff
-source: "https://raw.githubusercontent.com/NikolaLohinski/terraform-provider-freebox/main/examples/file-to-download.txt",
+source: "https://raw.githubusercontent.com/holyhope/terraform-provider-freebox/refs/heads/resources/remote_file/examples/provider.freebox.tf",
```

NikolaLohinski commented:
Hey @holyhope I just merged both PRs in free-go. Thanks for the contribution !

I'll cut a release in the morning (would have done it now but there's no way to tag from the web UI and I am AFK).

Ping me when you feel this MR is ready for review and we can go from there.

Awesome stuff already 👏🏻

NikolaLohinski commented:
https://github.com/NikolaLohinski/free-go/tree/v1.1.0 is out

holyhope commented:
Ready to review :-)

I delayed a few feature because they are not easy to implement (see the PR summary).
I suggest to do that in a second step.

NikolaLohinski commented on docs/resources/remote_file.md:
```diff
@@ -0,0 +1,56 @@
+# `freebox_remote_file` (Resource)
+
+Manages a virtual machine instance within a Freebox box. See the [Freebox blog](https://dev.freebox.fr/blog/?p=5450) for additional details
```
This looks like copy pasting from the VM definition

NikolaLohinski commented on internal/resource_remote_file.go:
```diff
@@ -0,0 +1,703 @@
+
+func (v *remoteFileResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "Manages a virtual machine instance within a Freebox box. See the [Freebox blog](https://dev.freebox.fr/blog/?p=5450) for additional details",
```
cf my previous comment

NikolaLohinski commented on internal/resource_remote_file.go:
```diff
@@ -0,0 +1,703 @@
+				Optional:           true,
+				Computed:           true,
+				MarkdownDescription: "Task identifier",
+				//PlanModifiers: []planmodifier.Int64{
+					// int64planmodifier.UseStateForUnknown(),
+				//},
```
```suggestion

```

NikolaLohinski commented on internal/resource_remote_file.go:
```diff
@@ -0,0 +1,703 @@
+
+// waitForFileSystemTask waits for the system task to complete.
+func waitForFileSystemTask(ctx context.Context, client client.Client, taskID int64) (error) {
+	tick := time.NewTicker(5 * time.Second)
```
That delay in between polls should probably be configurable at the resource level or something ; could help speed up the tests

Same goes for `waitForDownloadTask`.

NikolaLohinski reviewed (commented):
Minor comments, and the go.mod needs a bump, once everything's ready in free-go **[EDIT]** :point_right: https://github.com/NikolaLohinski/free-go/releases/tag/v1.2.0 is out

But all in all looks all very good to me.

Nice cleanup on the test setting up logic, thanks a lot for that

holyhope commented:
Note: I am not sure why the tests fail, it seems that some env variables are missing.

NikolaLohinski commented:
> Note: I am not sure why the tests fail, it seems that some env variables are missing.

Yeah it's because the testing workflow runs in a PR, which does not have access to the secrets which contain creds to my Freebox. It only works off of master branch.

There's probably better ways to do it but, coming from the GitLab world I am a noob in GitHub Actions 😂 

I'll have a local run today to get it merged, no worries.

NikolaLohinski commented:
Hmm I am running into issues when running it locally. Any clue ? :point_down: 
```
$ mage go:test
Running target: Go:Test
exec: ginkgo "run" "-p" "./..."
Running Suite: internal - /home/user/Documents/go/terraform-provider-freebox/internal
=======================================================================================
Random Seed: 1729235473

Will run 13 of 13 specs
Running in parallel across 11 processes
•
------------------------------
• [FAILED] [1.235 seconds]
resource "freebox_remote_file" { ... } create and delete (CD) without a checksum [It] should download and delete the file
/home/user/Documents/go/terraform-provider-freebox/internal/resource_remote_file_test.go:41

  [FAILED] Step 1/1 error: Error running apply: exit status 1

  Error: Value Conversion Error

    with freebox_remote_file.test-ac5d2b1b-0ff2-4f84-9f3d-4199be84de74,
  An unexpected error was encountered trying to build a value. This is always
  an error in the provider. Please report the following to the provider
  developer:

  Received unknown value, however the target type cannot handle unknown values.
  Use the corresponding `types` package type or a custom type that handles
  unknown values.

  Path: 
  Target Type: internal.pollingSpecModel
  Suggested Type: basetypes.ObjectValue

  Error: Provider returned invalid result object after apply

  After the apply operation, the provider still indicated an unknown value for
  freebox_remote_file.test-ac5d2b1b-0ff2-4f84-9f3d-4199be84de74.checksum. All
  values must be known after apply, so this is always a bug in the provider and
  should be reported in the provider's own repository. Terraform will still
  save the other known object values in the state.

  Error: Provider returned invalid result object after apply

  After the apply operation, the provider still indicated an unknown value for
  freebox_remote_file.test-ac5d2b1b-0ff2-4f84-9f3d-4199be84de74.polling.checksum_compute.
  All values must be known after apply, so this is always a bug in the provider
  and should be reported in the provider's own repository. Terraform will still
  save the other known object values in the state.

  Error: Provider returned invalid result object after apply

  After the apply operation, the provider still indicated an unknown value for
  freebox_remote_file.test-ac5d2b1b-0ff2-4f84-9f3d-4199be84de74.polling.create.
  All values must be known after apply, so this is always a bug in the provider
  and should be reported in the provider's own repository. Terraform will still
  save the other known object values in the state.

  Error: Provider returned invalid result object after apply

  After the apply operation, the provider still indicated an unknown value for
  freebox_remote_file.test-ac5d2b1b-0ff2-4f84-9f3d-4199be84de74.polling.delete.
  All values must be known after apply, so this is always a bug in the provider
  and should be reported in the provider's own repository. Terraform will still
  save the other known object values in the state.
  
  In [It] at: /home/user/Documents/go/terraform-provider-freebox/internal/resource_remote_file_test.go:42 @ 10/18/24 09:11:18.323
------------------------------
•
------------------------------
• [FAILED] [0.937 seconds]
resource "freebox_remote_file" { ... } create, update and delete (CUD) [It] should create, update and finally delete a file
/home/user/Documents/go/terraform-provider-freebox/internal/resource_remote_file_test.go:202

  [FAILED] Step 1/2 error: Error running apply: exit status 1

  Error: Value Conversion Error

    with freebox_remote_file.test-b68a1e51-6351-4238-9b74-11d9395c580f,
  An unexpected error was encountered trying to build a value. This is always
  an error in the provider. Please report the following to the provider
  developer:

  Received unknown value, however the target type cannot handle unknown values.
  Use the corresponding `types` package type or a custom type that handles
  unknown values.

  Path: 
  Target Type: internal.pollingSpecModel
  Suggested Type: basetypes.ObjectValue

  Error: Failed to get file

    with freebox_remote_file.test-b68a1e51-6351-4238-9b74-11d9395c580f,
    on terraform_plugin_test.tf line 16, in resource "freebox_remote_file" "test-b68a1e51-6351-4238-9b74-11d9395c580f":
    16: 							resource "freebox_remote_file" "test-b68a1e51-6351-4238-9b74-11d9395c580f" {

  path not found
  
  In [It] at: /home/user/Documents/go/terraform-provider-freebox/internal/resource_remote_file_test.go:203 @ 10/18/24 09:11:18.353
------------------------------
• [FAILED] [1.240 seconds]
resource "freebox_remote_file" { ... } create and delete (CD) with a checksum [It] should download, verify the checksum and delete the file
/home/user/Documents/go/terraform-provider-freebox/internal/resource_remote_file_test.go:107

  [FAILED] Step 1/1 error: Error running apply: exit status 1

  Error: Value Conversion Error

    with freebox_remote_file.test-29e646a7-ade1-46a3-9e3b-9e2675481686,
  An unexpected error was encountered trying to build a value. This is always
  an error in the provider. Please report the following to the provider
  developer:

  Received unknown value, however the target type cannot handle unknown values.
  Use the corresponding `types` package type or a custom type that handles
  unknown values.

  Path: 
  Target Type: internal.pollingSpecModel
  Suggested Type: basetypes.ObjectValue

  Error: Provider returned invalid result object after apply

  After the apply operation, the provider still indicated an unknown value for
  freebox_remote_file.test-29e646a7-ade1-46a3-9e3b-9e2675481686.polling.checksum_compute.
  All values must be known after apply, so this is always a bug in the provider
  and should be reported in the provider's own repository. Terraform will still
  save the other known object values in the state.

  Error: Provider returned invalid result object after apply

  After the apply operation, the provider still indicated an unknown value for
  freebox_remote_file.test-29e646a7-ade1-46a3-9e3b-9e2675481686.polling.create.
  All values must be known after apply, so this is always a bug in the provider
  and should be reported in the provider's own repository. Terraform will still
  save the other known object values in the state.

  Error: Provider returned invalid result object after apply

  After the apply operation, the provider still indicated an unknown value for
  freebox_remote_file.test-29e646a7-ade1-46a3-9e3b-9e2675481686.polling.delete.
  All values must be known after apply, so this is always a bug in the provider
  and should be reported in the provider's own repository. Terraform will still
  save the other known object values in the state.
  
  In [It] at: /home/user/Documents/go/terraform-provider-freebox/internal/resource_remote_file_test.go:108 @ 10/18/24 09:11:18.429
------------------------------
• [FAILED] [1.043 seconds]
resource "freebox_remote_file" { ... } create, update and delete (CUD) when only the task changed [It] should update the resource
/home/user/Documents/go/terraform-provider-freebox/internal/resource_remote_file_test.go:302

  [FAILED] Step 1/2 error: Error running apply: exit status 1

  Error: Value Conversion Error

    with freebox_remote_file.test-cd5b761e-316b-4c9e-aca6-de95b2cfab66,
  An unexpected error was encountered trying to build a value. This is always
  an error in the provider. Please report the following to the provider
  developer:

  Received unknown value, however the target type cannot handle unknown values.
  Use the corresponding `types` package type or a custom type that handles
  unknown values.

  Path: 
  Target Type: internal.pollingSpecModel
  Suggested Type: basetypes.ObjectValue

  Error: Provider returned invalid result object after apply

  After the apply operation, the provider still indicated an unknown value for
  freebox_remote_file.test-cd5b761e-316b-4c9e-aca6-de95b2cfab66.polling.checksum_compute.
  All values must be known after apply, so this is always a bug in the provider
  and should be reported in the provider's own repository. Terraform will still
  save the other known object values in the state.

  Error: Provider returned invalid result object after apply

  After the apply operation, the provider still indicated an unknown value for
  freebox_remote_file.test-cd5b761e-316b-4c9e-aca6-de95b2cfab66.polling.create.
  All values must be known after apply, so this is always a bug in the provider
  and should be reported in the provider's own repository. Terraform will still
  save the other known object values in the state.

  Error: Provider returned invalid result object after apply

  After the apply operation, the provider still indicated an unknown value for
  freebox_remote_file.test-cd5b761e-316b-4c9e-aca6-de95b2cfab66.polling.delete.
  All values must be known after apply, so this is always a bug in the provider
  and should be reported in the provider's own repository. Terraform will still
  save the other known object values in the state.
  
  In [It] at: /home/user/Documents/go/terraform-provider-freebox/internal/resource_remote_file_test.go:303 @ 10/18/24 09:11:18.553
------------------------------
• [FAILED] [0.702 seconds]
resource "freebox_virtual_machine" { ... } create and delete (CD) [It] should create, start, stop and delete a virtual machine
/home/user/Documents/go/terraform-provider-freebox/internal/resource_virtual_machine_test.go:19

  [FAILED] Step 1/1 error: Error running apply: exit status 1

  Error: Failed to start virtual machine

    with freebox_virtual_machine.test-CD-f5b62dfd-af07-4ef8,
    on terraform_plugin_test.tf line 16, in resource "freebox_virtual_machine" "test-CD-f5b62dfd-af07-4ef8":
    16: 							resource "freebox_virtual_machine" "test-CD-f5b62dfd-af07-4ef8" {

  failed to start virtual machine: failed to POST to vm/0/start endpoint:
  failed with error code 'file_not_found': Impossible de démarrer cette VM : Le
  fichier de disque n’a pas été trouvé
  
  In [It] at: /home/user/Documents/go/terraform-provider-freebox/internal/resource_virtual_machine_test.go:22 @ 10/18/24 09:11:18.871
------------------------------
SS••••

Summarizing 5 Failures:
  [FAIL] resource "freebox_remote_file" { ... } create and delete (CD) without a checksum [It] should download and delete the file
  /home/user/Documents/go/terraform-provider-freebox/internal/resource_remote_file_test.go:42
  [FAIL] resource "freebox_remote_file" { ... } create, update and delete (CUD) [It] should create, update and finally delete a file
  /home/user/Documents/go/terraform-provider-freebox/internal/resource_remote_file_test.go:203
  [FAIL] resource "freebox_remote_file" { ... } create and delete (CD) with a checksum [It] should download, verify the checksum and delete the file
  /home/user/Documents/go/terraform-provider-freebox/internal/resource_remote_file_test.go:108
  [FAIL] resource "freebox_remote_file" { ... } create, update and delete (CUD) when only the task changed [It] should update the resource
  /home/user/Documents/go/terraform-provider-freebox/internal/resource_remote_file_test.go:303
  [FAIL] resource "freebox_virtual_machine" { ... } create and delete (CD) [It] should create, start, stop and delete a virtual machine
  /home/user/Documents/go/terraform-provider-freebox/internal/resource_virtual_machine_test.go:22

Ran 11 of 13 Specs in 4.095 seconds
FAIL! -- 6 Passed | 5 Failed | 0 Pending | 2 Skipped

```

holyhope commented:
> @holyhope I am running into issues when running it locally. Any clue ? 👇
> 
> [...]
> 
> FAIL! -- 6 Passed | 5 Failed | 0 Pending | 2 Skipped
> ```

I fixed the issue by adding some default values.
I added some tests to cover these cases (It might not cover all of them).

Can you tell me it fixes the issue on your side too please?

NikolaLohinski commented on internal/resource_virtual_machine_test.go:
```diff
@@ -121,12 +51,22 @@ var _ = Context("resource \"freebox_virtual_machine\" { ... }", Ordered, func()
+				CheckDestroy: func(s *terraform.State) error {
+					id, err := strconv.Atoi(s.RootModule().Resources["freebox_virtual_machine."+name].Primary.Attributes["id"])
+					Expect(err).To(BeNil())
+					Expect(id).ToNot(BeZero())
```
Could be the first VM ever. Checking for 0 value here is erroneous
```suggestion
```
